rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+
+## Why
+
+When Teleport Kube Agent wants to join a Teleport Cluster, it needs to share the invite token with the cluster for initial authentication. The invite token can be:
+
+- short-lived token (the token will expire after a low TTL and cannot be reused after that time)
+- long-lived/static token (the usage of long-lived tokens is discouraging for security reasons).
+
+After sharing the invite token with the Teleport cluster, the agent receives its identity from the Auth service. Identity certificates are mandatory for accessing the Teleport Cluster and must be stored for accesses without reusing the invite token.
+
+Kubernetes Pods are, by definition, expected to be stateless. This means that each time a Pod is recycled because it was restarted, deleted, upgraded, or moved to another node, the state that was written to its filesystem is lost.
+
+One way to overcome this problem is to use Persistent Volumes. PV is a Kubernetes feature that mounts a storage volume in the container filesystem, whose lifecycle is independent of the Pod that mounts it. Kubernetes’ PV storage has its own drawbacks. Persistent Volumes are very distinct between different cloud vendors. Some providers lock volumes to be mounted in the same zone they were created, meaning that if Teleport agent Pod is recycled, it must be maintained in the same zone it was created. This creates operational issues and for on-premises deployments might be difficult to manage.
+
+Another possibility is that the Kube Agent might use the invite token each time the pod is recycled, meaning that it issues a join request to the cluster every time it starts and receives a new identity from the Auth service. This means that the invite token must be a static/long-lived token, otherwise after a while the agent could not register himself in the cluster because the invite token expired. This approach is not recommended and might cause security flaws because the join token can be stolen or guessed.
+
+One solution that might address all the issues referenced above is to allow the Kube agent to use Kubernetes secrets as storage backend. This allows not only that the agent is able to run stateless depending only on native objects generally available in any Kubernetes cluster, but also that the agent might support dynamic short-lived invite tokens with no dependency on external storage.
+
+## Details
+
+### Limitations
+
+High availability can only be achieved using Statefulsets and not deployments. This means that if the desired number of replicas is bigger than one, i.e `replicas>1`, the [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) has to switch to Statefulset objects instead of Deployments. This change is required because some Kubernetes invariant must be kept between restarts in order to correctly map each agent pod to its identity stored in the secret. The invariant used is the Statefulset pod names `{{ .Release.Name }}-{0...replicas}}`.
```

Comment:
How do we handle migration from Deployment to StatefulSet when customers upgrade their Helm chart? Might be worth adding a section on that.